Hook up IPC for registered scripts APIs.

kiaraarose opened the pull request:
#### 27ba8dee1e35d5d7a700b78dbabf8f7b3b0cafcf
<pre>
Hook up IPC for registered scripts APIs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=264823">https://bugs.webkit.org/show_bug.cgi?id=264823</a>
<a href="https://rdar.apple.com/118405149">rdar://118405149</a>

Reviewed by Timothy Hatcher.

This patch creates a WebExtensionRegisteredScriptParameters which will be used to pass information
to the UI process for the scripts to be registered for scripting.RegisteredContentScripts. Also adds
tests for error handling for these APIs in WKWebExtensionAPIScripting.Errors.

This patch also restructures the WebExtensionAPIScripting class to match that of the other classes
created for other extension APIs.

* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/Extensions/WebExtensionDynamicScripts.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptParameters.h: Copied from Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::WebExtensionContext::scriptingRegisterScripts):
(WebKit::WebExtensionContext::scriptingUpdateRegisteredScripts):
(WebKit::WebExtensionContext::scriptingGetRegisteredScripts):
(WebKit::WebExtensionContext::scriptingUnregisterScripts):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::toWebAPI):
(WebKit::WebExtensionAPIScripting::executeScript):
(WebKit::WebExtensionAPIScripting::insertCSS):
(WebKit::WebExtensionAPIScripting::removeCSS):
(WebKit::WebExtensionAPIScripting::registerContentScripts):
(WebKit::WebExtensionAPIScripting::getRegisteredContentScripts):
(WebKit::WebExtensionAPIScripting::updateContentScripts):
(WebKit::WebExtensionAPIScripting::unregisterContentScripts):
(WebKit::WebExtensionAPIScripting::validateScript):
(WebKit::WebExtensionAPIScripting::validateRegisteredScripts):
(WebKit::WebExtensionAPIScripting::validateFilter):
(WebKit::WebExtensionAPIScripting::parseTargetInjectionOptions):
(WebKit::WebExtensionAPIScripting::parseScriptInjectionOptions):
(WebKit::WebExtensionAPIScripting::parseCSSInjectionOptions):
(WebKit::WebExtensionAPIScripting::parseRegisteredContentScripts):
(WebKit::WebExtensionAPIScripting::hasValidExecutionWorld):
(WebKit::parseTargetInjectionOptions): Deleted.
(WebKit::parseScriptInjectionOptions): Deleted.
(WebKit::parseCSSInjectionOptions): Deleted.
(WebKit::WebExtensionAPIScripting::toWebAPI): Deleted.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::executeScript):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/270733@main">https://commits.webkit.org/270733@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/408cced01f4ff9ee62cbc9f018aac52721a64661

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26277 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/4886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/27538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/28372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/24051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/26609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6669 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/2299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/28372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26533 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/3730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/27538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/28951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/26440 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/27538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/28951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/23947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/27538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/28951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/3363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/2299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/4786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/27538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/3841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3381 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/3688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -141,6 +142,35 @@
     completionHandler(std::nullopt);
 }
 
+void WebExtensionContext::scriptingRegisterScripts(const Vector<WebExtensionRegisteredScriptParameters> &parameters, CompletionHandler<void(WebExtensionDynamicScripts::Error)>&& completionHandler)
```
`const Vector<WebExtensionRegisteredScriptParameters>& ` instead and the same for the methods below

kiaraarose replied:
`const Vector<WebExtensionRegisteredScriptParameters>& scripts`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -59,81 +61,67 @@
-    }
+    for (auto& parameters : parametersVector) {
+        NSMutableDictionary *result = [NSMutableDictionary dictionaryWithCapacity:3];
 
```
`auto *result = [NSMutableDictionary dictionaryWithCapacity:3];`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -349,29 +379,184 @@ static inline void parseCSSInjectionOptions(NSDictionary *cssInfo, WebExtensionS
+        }
+
+        if (NSString *injectionTime = script[runAtKey]) {
+            if (![injectionTime isEqualToString:documentIdle] || ![injectionTime isEqualToString:documentStart] || ![injectionTime isEqualToString:documentStart]) {
```
`if (![injectionTime isEqualToString:documentIdle] || ![injectionTime isEqualToString:documentStart] || ![injectionTime isEqualToString:documentEnd)`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -349,29 +379,184 @@ static inline void parseCSSInjectionOptions(NSDictionary *cssInfo, WebExtensionS
+void WebExtensionAPIScripting::parseTargetInjectionOptions(NSDictionary *targetInfo, WebExtensionScriptInjectionParameters& parameters, NSString **outExceptionString)
+{
+    NSNumber *tabID = targetInfo[tabIDKey];
+    std::optional<WebExtensionTabIdentifier> identifier = toWebExtensionTabIdentifier(tabID.doubleValue);
```
`auto tabIdentifier = toWebExtensionTabIdentifier(tabID.doubleValue);`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -349,29 +379,184 @@ static inline void parseCSSInjectionOptions(NSDictionary *cssInfo, WebExtensionS
+    if (NSArray *frameIDs = targetInfo[frameIDsKey]) {
+        Vector<WebExtensionFrameIdentifier> frames;
+        for (NSNumber *frameID in frameIDs) {
+            auto identifier = toWebExtensionFrameIdentifier(frameID.doubleValue);
```
`frameIdentifier`

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionDynamicScripts.serialization.in:
```diff
@@ -44,4 +44,21 @@ struct WebKit::WebExtensionScriptInjectionResultParameters {
+    std::optional<Vector<String>> js;
+
+    String identifier;
+    String runAt;
```
This should use `WebCore::UserScriptInjectionTime` or `WebExtension::InjectionTime` enum (likely the WebExtension enum). We want to avoid leaking the API naming, and parsing, up into the UI implementation where possible.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -59,81 +61,67 @@
-    if (NSArray *files = script[filesKey])
-        parameters.files = makeVector<String>(files);
+        if (parameters.result)
+            result[@"result"] = (NSString *)parameters.result.value();
```
I don't know if the cast is needed. I wasn't seeing error when doing this without the cast in one of @b-weinstein's PRs, so it might Just Work™.

kiaraarose replied:
yea i don't think it's needed. i added it based on the feedback from the last PR but will revert back

xeenon replied:
Yeah, I have thought it was needed in this case but if it works without, great!

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -349,29 +379,184 @@ static inline void parseCSSInjectionOptions(NSDictionary *cssInfo, WebExtensionS
+            return false;
+        }
+
+        if (isRegisteringScript && !((NSArray *)script[cssKey]).count && !((NSArray *)script[jsKey]).count) {
```
Putting these in local vars might be cleaner than the casts, similar to `matchPatterns` and `scriptID`.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -141,6 +142,35 @@
+    completionHandler({ }, std::nullopt);
+}
+
+
```
remove extra newline

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -349,29 +379,184 @@ static inline void parseCSSInjectionOptions(NSDictionary *cssInfo, WebExtensionS
+void WebExtensionAPIScripting::parseScriptInjectionOptions(NSDictionary *script, WebExtensionScriptInjectionParameters& parameters)
+{
+    if (script[funcKey] || script[functionKey]) {
+        NSString *key = script[funcKey] ? funcKey : functionKey;
```
`auto *`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -349,29 +379,184 @@ static inline void parseCSSInjectionOptions(NSDictionary *cssInfo, WebExtensionS
+    }
+
+    if (script[argsKey] || script[argumentsKey]) {
+        NSString *key = script[argsKey] ? argsKey : argumentsKey;
```
`auto *`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -349,29 +379,184 @@ static inline void parseCSSInjectionOptions(NSDictionary *cssInfo, WebExtensionS
+    if (NSArray *files = script[filesKey])
+        parameters.files = makeVector<String>(files);
+
+    if ([(NSString *)script[worldKey] isEqualToString:mainWorld])
```
Cast shouldn't be needed?

kiaraarose committed:
Hook up IPC for registered scripts APIs.
https://bugs.webkit.org/show_bug.cgi?id=264823
rdar://118405149

Reviewed by Timothy Hatcher.

This patch creates a WebExtensionRegisteredScriptParameters which will be used to pass information
to the UI process for the scripts to be registered for scripting.RegisteredContentScripts. Also adds
tests for error handling for these APIs in WKWebExtensionAPIScripting.Errors.

This patch also restructures the WebExtensionAPIScripting class to match that of the other classes
created for other extension APIs.

* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/Extensions/WebExtensionDynamicScripts.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptParameters.h: Copied from Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::WebExtensionContext::scriptingRegisterScripts):
(WebKit::WebExtensionContext::scriptingUpdateRegisteredScripts):
(WebKit::WebExtensionContext::scriptingGetRegisteredScripts):
(WebKit::WebExtensionContext::scriptingUnregisterScripts):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::toWebAPI):
(WebKit::WebExtensionAPIScripting::executeScript):
(WebKit::WebExtensionAPIScripting::insertCSS):
(WebKit::WebExtensionAPIScripting::removeCSS):
(WebKit::WebExtensionAPIScripting::registerContentScripts):
(WebKit::WebExtensionAPIScripting::getRegisteredContentScripts):
(WebKit::WebExtensionAPIScripting::updateContentScripts):
(WebKit::WebExtensionAPIScripting::unregisterContentScripts):
(WebKit::WebExtensionAPIScripting::validateScript):
(WebKit::WebExtensionAPIScripting::validateRegisteredScripts):
(WebKit::WebExtensionAPIScripting::validateFilter):
(WebKit::WebExtensionAPIScripting::parseTargetInjectionOptions):
(WebKit::WebExtensionAPIScripting::parseScriptInjectionOptions):
(WebKit::WebExtensionAPIScripting::parseCSSInjectionOptions):
(WebKit::WebExtensionAPIScripting::parseRegisteredContentScripts):
(WebKit::WebExtensionAPIScripting::hasValidExecutionWorld):
(WebKit::parseTargetInjectionOptions): Deleted.
(WebKit::parseScriptInjectionOptions): Deleted.
(WebKit::parseCSSInjectionOptions): Deleted.
(WebKit::WebExtensionAPIScripting::toWebAPI): Deleted.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::executeScript):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/270733@main
